### PR TITLE
Parse payload template at initialize not runtime

### PR DESCRIPTION
**Pull request by agrare**
Similar to https://github.com/ManageIQ/floe/pull/71

Parse PayloadTemplate values at workflow parse time rather than runtime.

### CODE REVIEW

File: lib/floe/workflow/payload_template.rb [hunk 1]
```diff
@@ -4,31 +4,56 @@ module Floe
   class Workflow
     class PayloadTemplate
       def initialize(payload)
-        @payload = payload
+        @payload = parse_payload(payload)
```
Comment (kbrock) [inline lib/floe/workflow/payload_template.rb:7]: I think of `payload` as the unparsed input.
(I also kinda of want to get rid of all unparsed payloads that we store)

could we call this something like `parameters` or `template`?

File: lib/floe/workflow/payload_template.rb [hunk 2]
```diff
@@ -4,31 +4,56 @@ module Floe
...
-        interpolate_value_nested(payload, context, inputs)
+        interpolate_value(payload, context, inputs)
       end
 
       private
 
       attr_reader :payload
 
-      def interpolate_value_nested(value, context, inputs)
+      def parse_payload(value)
```
Comment (kbrock) [inline lib/floe/workflow/payload_template.rb:18]: so cool to get `context` out of the parsing function. 🎉

File: lib/floe/workflow/payload_template.rb [hunk 3]
```diff
@@ -4,31 +4,56 @@ module Floe
...
+            Path.new(value)
+          else
+            value
+          end
+        else
+          value
+        end
+      end
+
+      def interpolate_value(value, context, inputs)
```
Comment (kbrock) [inline lib/floe/workflow/payload_template.rb:43]: this works as you described it. And maybe we want to merge this as is and plan a possible continuation?

Does it make sense to take it a little further and have `PayloadTemplateArray`, `PayloadTemplateHash`, `PayloadTemplateStatic`, and `Path`?

Then all methods are like a command pattern with:

```ruby
def value(context, inputs)
  @template.value(context, inputs)
end
```

Maybe the recursive nature breaks this down too quickly.
Think I do like keeping `Path`, `.$` and static values hidden from the array/hash/constant parser.

Not sure if this is overcomplicating things though.

File: spec/workflow/payload_template_spec.rb [hunk 4]
```diff
@@ -19,6 +19,24 @@
       it "returns the value from the inputs" do
         expect(subject.value(context, inputs)).to eq({"foo" => "bar", "bar" => "baz"})
       end
+
+      context "with an invalid payload" do
+        let(:payload) { {"foo.$" => "$.foo", "foo" => "bar"} }
+
+        it "raises an exception" do
+          expect { subject }.to raise_error(Floe::InvalidWorkflowError, "both foo.$ and foo present")
```
Comment (kbrock) [inline spec/workflow/payload_template_spec.rb:27]: nice

**Review by kbrock — COMMENTED** [review]
This is going to be very nice.
Just those 2 comments

File: lib/floe/workflow/payload_template.rb (hunk 1 above)
Comment (agrare) [reply lib/floe/workflow/payload_template.rb:7]: 👍 yeah good point

File: spec/workflow/payload_template_spec.rb (hunk 4 above)
Comment (agrare) [reply spec/workflow/payload_template_spec.rb:27]: Yeah I noticed this in the spec and realized it was a super easy validation to add now

**Review by Fryguy — APPROVED** [review]

File: lib/floe/workflow/payload_template.rb [hunk 5]
```diff
@@ -4,35 +4,73 @@ module Floe
   class Workflow
     class PayloadTemplate
       def initialize(payload)
-        @payload = payload
+        @payload          = payload
```
Comment (Fryguy) [inline lib/floe/workflow/payload_template.rb:7]: Hard to tell from this PR, but do we even need `@payload` anymore,  or is just having `@payload_template` sufficient?

Comment (agrare) [reply lib/floe/workflow/payload_template.rb:7]: It isn't used, we could remove it I just had it for consistency with the other classes which keep the raw payload

Comment (agrare) [reply lib/floe/workflow/payload_template.rb]: Removed

**Comment by Fryguy** [comment]
I agree with the bot to avoid pinging `@payload` in the commit message (poor guy)

File: lib/floe/workflow/payload_template.rb (hunk 5 above)
Comment (kbrock) [reply lib/floe/workflow/payload_template.rb]: Yay. I want to remove payload (and workflow) from everywhere. But more of a side thought and don't have any good push behind it

File: lib/floe/workflow/payload_template.rb [hunk 6]
```diff
@@ -4,35 +4,85 @@ module Floe
   class Workflow
     class PayloadTemplate
       def initialize(payload)
-        @payload = payload
+        verify_payload!(payload)
```
Comment (kbrock) [inline lib/floe/workflow/payload_template.rb:7]: can we incorporate this in `parse_payload`?
Or do you feel it works better if you have separate?

I feel like an `else` or something in `parse` may meet our needs

Comment (agrare) [reply lib/floe/workflow/payload_template.rb:7]: :laughing: I moved this out of there to appease codeclimate.  I actually agree it was a little simpler in `parse_payload` but that cognitive complexity was too high

File: spec/workflow/payload_template_spec.rb (hunk 4 above)
Comment (kbrock) [reply spec/workflow/payload_template_spec.rb:27]: aah. here is a good example of where a separate validation step is probably better than a parse step. But still curious your take.

File: lib/floe/workflow/payload_template.rb (hunk 6 above)
Comment (agrare) [reply lib/floe/workflow/payload_template.rb]: Okay you convinced me to not change code just to fix a 6/5 cyclomatic complexity

**Pull request merged**